lib/ecto/association.ex

```diff
@@ -261,6 +261,8 @@ defmodule Ecto.Association do
   """
   def related_from_query(atom) when is_atom(atom), do: atom
   def related_from_query({source, schema}) when is_binary(source) and is_atom(schema), do: schema
+  def related_from_query(%Ecto.Query{from: schema}) when is_atom(schema), do: schema
+  def related_from_query(%Ecto.Query{from: {source, schema}}) when is_binary(source) and is_atom(schema), do: schema
```

Comment:
We don't need two clauses here because queries are always in the form `from: {source, schema}`. However, we do need to check that it only has a `where` field and nothing more. Something like this:

```elixir
case query do
      %Ecto.Query{order_bys: [], limit: nil, offset: nil, group_bys: [], joins: [],
                  havings: [], preloads: [], assocs: [], distinct: nil, lock: nil} ->
        query
      _ ->
        raise ...
    end
```